### PR TITLE
Modernize it

### PR DESCRIPTION
**Pull request by filosganga**
This upgrade the profiles service to the latest comms standards:
 - Upgrade Http4s to 0.18.x
 - Drop typesafe config in favour of ciris
 - Add automatic docker image versioning
 - Drop docker-compose based tests in favour of DokerKit
 - Add circleci 2.0 workflow with git tagging and e2e tests

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -1,25 +0,0 @@
-version: '2'
```
Comment (laurence-bird) [inline docker-compose.yml]: 🎉

File: get_ip_address.sh [hunk 2]
```diff
@@ -1,5 +0,0 @@
-#!/usr/bin/env bash
-# Get the first non-loopback address of the machine
-
-export IP_ADDRESS="$(ifconfig | grep -Eo 'inet (addr:)?([0-9]*\.){3}[0-9]*' | grep -Eo '([0-9]*\.){3}[0-9]*' | grep -v '127.0.0.1' | head -n 1)"
```
Comment (laurence-bird) [inline get_ip_address.sh]: 🎉 🎉

File: src/main/scala/com/ovoenergy/comms/profiles/Config.scala [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+    case other => Left(s"$other is not a known environment")
+  }
+
+}
+
+object Config {
+
+  def load[F[_]: Sync]: F[Config] = {
+
+    implicit val cirisSyncForF: ciris.api.Sync[F] = ciris.cats.effect.catsEffectSyncToCiris(Sync[F])
```
Comment (SystemFw) [inline src/main/scala/com/ovoenergy/comms/profiles/Config.scala]: Why is this needed explicitly?

File: src/main/scala/com/ovoenergy/comms/profiles/Http.scala [hunk 4]
```diff
@@ -37,14 +42,13 @@ object Http extends LoggingWithCustomerId {
...
-              s"Failed to retrieve customer profile."
+            val errorStr             = s"Failed to retrieve customer profile."
             logError(customerId)(s"$errorStr, Error: ${err.message} SF Response: $statusCodeLogMessage")
             InternalServerError(errorStr)
           }
           case Right(customer) => {
             ProfileResponse.fromCustomer(customer) match {
-              case Right(r) => Ok(r.asJson.spaces2)
+              case Right(r) => Ok(r.asJson)
               case Left(notFoundErr) => {
```
Comment (SystemFw) [inline src/main/scala/com/ovoenergy/comms/profiles/Http.scala]: Not that big of a deal now, since this is a blocker, but all this pattern matching with `Left` and `Right` can potentially be improved by some monaderror

File: src/main/scala/com/ovoenergy/comms/profiles/Http.scala [hunk 5]
```diff
@@ -56,22 +60,29 @@ object Http extends LoggingWithCustomerId {
...
-        else left("Invalid API key")
+    import cats.implicits._
+
+    val checkApiKey: Kleisli[F, Request[F], Either[String, Unit]] = Kleisli { request =>
+      apiKeyValidation(request).map { isValid =>
+        if (isValid) {
+          Right(())
+        } else {
+          Left("Invalid API key")
+        }
```
Comment (SystemFw) [inline src/main/scala/com/ovoenergy/comms/profiles/Http.scala]: potentially (untested)

```scala
apiKeyValidationRequest(request).map(_.guard[Option].toRight("Invalid API key"))

File: src/main/scala/com/ovoenergy/comms/profiles/Main.scala [hunk 6]
```diff
@@ -1,70 +1,65 @@
...
+          emailAddress = Some(primaryEmailAddress),
+          phoneNumber = Some(config.canary.mobileNumber),
+          List(ResponseCommunicationPreference(Service, List(Email, SMS)),
+               ResponseCommunicationPreference(Marketing, List(Email, SMS)))
+        )
+      }
 
-  Files.readAllLines(Paths.get("banner.txt")).asScala.foreach(println)
+      BlazeBuilder[IO]
+        .withBanner(Source.fromResource("banner.txt").getLines().toList)
```
Comment (laurence-bird) [inline src/main/scala/com/ovoenergy/comms/profiles/Main.scala]: sweet feature

Comment (SystemFw) [reply src/main/scala/com/ovoenergy/comms/profiles/Main.scala]: You're forgiven for now, but do it with fs2 next time :P

File: src/main/scala/com/ovoenergy/comms/profiles/Config.scala (hunk 3 above)
Comment (filosganga) [reply src/main/scala/com/ovoenergy/comms/profiles/Config.scala]: It should not be needed, but I believe there is a conflict with implicits that make it fail to resolve.

File: src/main/scala/com/ovoenergy/comms/profiles/Http.scala (hunk 4 above)
Comment (filosganga) [reply src/main/scala/com/ovoenergy/comms/profiles/Http.scala]: Agree, I find myself using a lot of pattern matching instead of combinators.

File: src/main/scala/com/ovoenergy/comms/profiles/Http.scala (hunk 5 above)
Comment (filosganga) [reply src/main/scala/com/ovoenergy/comms/profiles/Http.scala]: I will try it!

File: src/main/scala/com/ovoenergy/comms/profiles/Main.scala (hunk 6 above)
Comment (filosganga) [reply src/main/scala/com/ovoenergy/comms/profiles/Main.scala]: LOL, will it work?

```
      fs2.io.file
        .readAll[IO](Paths.get(ClassLoader.getSystemClassLoader.getResource("banner.txt").toURI), 1024)
        .through(fs2.text.utf8Decode)
        .compile
        .fold("")(_ ++ _)
        .unsafeRunSync()
```

Comment (SystemFw) [reply src/main/scala/com/ovoenergy/comms/profiles/Main.scala]: it's fine, you can leave it as is since it's just a one liner (btw you can do `through(text.lines).compile.toList`) but whatever

Comment (filosganga) [reply src/main/scala/com/ovoenergy/comms/profiles/Main.scala]: Will do in a next iteration, thanks
